tutorials/early_stopping/early_stopping.ipynb

```diff
@@ -0,0 +1,978 @@
+   "id": "57aa9cf7",
+   "metadata": {},
+   "source": [
+    "## Defining an Early Stopping Strategy\n",
+    "\n",
+    "A `PercentileEarlyStoppingStrategy` is a simple method that stops a trial if its performance falls below a certain percentile of other trials at the same step (e.g., when `percentile_threshold` is 50, at a given point in time, if a trial ranks in the bottom 50% of trials, it is stopped). \n",
+    "- We make use of `normalize_progressions` which normalizes the progression column (e.g. timestamp, epochs, training data used) to be in [0, 1]. This is useful because one doesn't need to know the maximum progression values of the curve (which might be, e.g., the total number of data points in the training dataset).\n",
+    "- The `min_progression` parameter specifies that trials should only be considered for stopping if the latest progression value is greater than this threshold.\n",
+    "- The `min_curves` parameter specifies the minimum number of completed curves before early stopping will be considered. This should be larger than zero if `normalize_progression` is used. In general, we want a few completed curves to have a baseline for comparison."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 8,
+   "id": "949e8ab5",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "percentile_early_stopping_strategy = PercentileEarlyStoppingStrategy(\n",
+    "    # stop if in bottom 50% of runs at the same progression\n",
+    "    percentile_threshold=50,\n",
+    "    # the trial must have passed `min_progression` steps before early stopping is initiated\n",
+    "    # note that we are using `normalize_progressions`, so this is on a scale of [0, 1]\n",
+    "    min_progression=0.3,\n",
+    "    # there must be `min_curves` completed trials and `min_curves` trials reporting data in\n",
+    "    # order for early stopping to be applicable\n",
+    "    min_curves=3,\n",
+    "    # specify, e.g., [0, 1] if the first two trials should never be stopped\n",
+    "    trial_indices_to_ignore=None,\n",
+    "    # check for new data every 30 seconds\n",
+    "    seconds_between_polls=30,\n",
+    "    normalize_progressions=True,\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "2665ca93",
+   "metadata": {},
+   "source": [
+    "## Creating the Ax Experiment\n",
+    "\n",
+    "In Ax, the Experiment object is the object that stores all the information about the problem setup."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 9,
+   "id": "12849b31",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "experiment = Experiment(\n",
+    "    name=\"torchx_mnist\",\n",
+    "    search_space=search_space,\n",
+    "    optimization_config=opt_config,\n",
+    "    runner=ax_runner,\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "49a4ed0e",
+   "metadata": {},
+   "source": [
+    "## Choosing the GenerationStrategy\n",
```

Comment:
It may make sense to acknowledge here that "proper" decision theoretic handling would treat generation and early stopping strategy jointly in one strategy, and that the separation here is for simplicity (and b/c it generally works quite well in practice).